plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/LazyConnectionFactory.java

```diff
@@ -32,7 +32,7 @@ public final class LazyConnectionFactory
     private final ConnectionFactory delegate;
 
     @Inject
-    public LazyConnectionFactory(@ForLazyConnectionFactory ConnectionFactory delegate)
+    public LazyConnectionFactory(@StatsCollecting ConnectionFactory delegate)
```

Comment:
Why not to bind `StatsCollectingConnectionFactory` directly? 

Surely it is not the question I want to ask. I wanted put your attention on the fact that now you hardcode how things should be injected in class implementation, whereas previously it was done in guice and classes looked independently. So potentially you could change the order of injection without changing class implementation.

I am not saying it is bad. I am just sharing concerns. If you think I am right then maybe we should not have `@StatsCollecting` annotation but something like `@For...`

If you think, my concern is an overkill because we want to hardcode dependency injection in the implementation of the class, then I think we should use `StatsCollectingConnectionFactory` directly here and remove `@StatsCollecting`. That would be coherent with your next commit where you inject implementation to classes (not interface).

Reply:
To answer the question you DID want to ask:

This is my interpretation of the other comment in this PR from @hashhar: https://github.com/trinodb/trino/pull/19191#discussion_r1344016312

> Readability. Our code is written for the readability of the next person

I've spent enough time trying to understand how this chain of responsibility is constructed, to learn that it actually looks like abusing YAGNI (https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) principle:
> So potentially you could change the order of injection without changing class implementation...

To answer the question you did not wanted to ask:

I wanted to bind `StatsCollectingConnectionFactory` directly, but then I noticed this in `JdbcDiagnosticModule`:
```
        newExporter(binder).export(Key.get(ConnectionFactory.class, StatsCollecting.class))
                .as(generator -> generator.generatedNameOf(ConnectionFactory.class, catalogName.get().toString()));
```
and I wasn't sure whether this means that `StatsCollecting` is also used externally. I wanted to stay on the safe side.

---

To summarize. I can go one way or the other. Whatever you tell me makes more sense. I don't have enough gut feeling here which is the best solution.

Reply:
@kokosing  - added this commit https://github.com/trinodb/trino/pull/19191/commits/14167f5b8e3efcc43d77bafeca6f9ff916db9a2a  'on top' as a try-out for simplifying even further. I didn't want to blend it into the other commits, as I'm still not sure whether you won't change your mind and then it's easier to remove the changes. Let me know.

Reply:
```
newExporter(binder).export(Key.get(ConnectionFactory.class, StatsCollecting.class))
                .as(generator -> generator.generatedNameOf(ConnectionFactory.class, catalogName.get().toString()));
```

This exposes an MBean in JMX so `@Managed` things are reported there. Then using JMX connector you can query these JMX stats.

Reply:
I am fine with the last commit. Let's squash it with `Simplify ConnectionFactory chain of responsibility` commit.